### ISSUE TITLE
2.0 release?

### ISSUE THREAD

**Issue by swails** [ParmEd/ParmEd#169]
What do people think about an official release?  There are obviously still lots of things worth adding, but there hasn't been an official release in a very long time.

I cut a beta pre-release in order to put out a conda package for people that needed one, but a real release is warranted, IMO.  I wanted to wait for the gromber command, but now that's done.

**Comment by kyleabeauchamp** [comment]
+1, maybe just wait until someone has tested out the conda packages to make sure there aren't any fixes needed there.

**Comment by swails** [comment]
Sounds good -- @jchodera, have you used the conda packages yet?  And if so, do they work for you?

I'll also try them.

**Comment by jchodera** [comment]
I was able to `conda install parmed` on my `osx` machine, but I don't actually know how to *import* it in Python to check which version number I got.  Apologies, but I'm a total n00b here.

I see `parmed.py` on my path (which is it not just called `parmed` if it is an executable?) though.

**Comment by davidlmobley** [comment]
You want 'chemistry'



—

**Comment by swails** [comment]
The version number is wrong because I forgot to change it.  It will say something like "15.1", unfortunately.  I'm thinking about moving to versioneer on @rmcgibbo's suggestion.

But to do that:

```Python
import chemistry as chem
chem.__version__
```

I'm also thinking about adding a version module that gives more information (kind of like what numpy and Python itself have).  The main motivation here is that ParmEd has its AmberTools versioning and its standalone versioning (yuck).  It's currently a mess, and I don't want any version ambiguity now that more people are starting to use it.

**Comment by swails** [comment]
But more than simply installing it, if you plan on using it on Travis that would be more of a stress-test.

**Comment by jchodera** [comment]
That works!
```
>>> import chemistry
>>> chemistry.__version__
'15.1'
```
Some suggestions (though I am sure you have heard them before):
* It would be useful to have your package name match your import name.  `conda install parmed` to do `import chemistry` makes my brain hurt.
* The version numbers should probably match too.  I'm very surprised that installing version 2.0beta0 yields version 15.1.

**Comment by jchodera** [comment]
> But more than simply installing it, if you plan on using it on Travis that would be more of a stress-test.

@ChayaSt is in the process of setting up travis-ci for her [torsion parameterization tool](https://github.com/choderalab/Torsions), so she'll use `parmed` for that!

**Comment by swails** [comment]
> It would be useful to have your package name match your import name. conda install parmed to do import chemistry makes my brain hurt.

See #85.  This is a hard decision for me.  It's been `chemistry` and `ParmedTools` too long for me to simply up and change it now.  Too many places rely on those imports to move them inside a `parmed` namespace (and out of the global import namespace).  In any case, #85 is where this discussion should take place.

> The version numbers should probably match too. I'm very surprised that installing version 2.0beta0 yields version 15.1.

As I said in my previous comment, the version number is wrong (it reads 15.1 instead of 2.0beta), because I forgot to change it before cutting the release (which was only done for conda availability).  Alas, I've never been good at remembering to update versioning information... sigh.

> @ChayaSt is in the process of setting up travis-ci for her torsion parameterization tool, so she'll use parmed for that!

Great.  Let me know how it goes.

**Comment by davidlmobley** [comment]
We can use it on Travis in the near future but it will take a couple of days probably.





—

**Comment by jchodera** [comment]
>  In any case, #85 is where this discussion should take place.

I might have poorly phrased my point here: Whatever you call it (I honestly don't have any opinions here!) your conda package name should be the same.  It's currently not.

**Comment by swails** [comment]
> I might have poorly phrased my point here: Whatever you call it (I honestly don't have any opinions here!) your conda package name should be the same. It's currently not.

ParmEd installs 2 top-level packages: `chemistry` and `ParmedTools` (and `fortranformat`, but that's an implementation detail that users will never use, so I might move that into `chemistry.utils`).  It also installs 2 scripts/programs: `parmed.py` and `xparmed.py` into the bin directory.  So the conda package name is the same as the *programs* that get installed and quite similar to one of the packages that gets installed (but obviously different from the main "workhorse" package).

In light of this, any package name would have drawbacks.

**Comment by jchodera** [comment]
Why not split this into two packages if they are logically independent?

**Comment by swails** [comment]
They're not logically independent.

**Comment by jchodera** [comment]
Sorry to be late to the game here: why are they in separate namespaces then?  If one is dependent on the other, they can still be split into separate conda packages and the dependencies can be encoded in the conda recipes.

If they are interdependent....well, that sounds more like a design flaw?

**Comment by leeping** [comment]
Just butting in here, but I realized fairly quickly (as I'm sure you did) that I needed to import `chemistry` for the functionality I needed rather than `ParmEd`.  What's wrong with Jason organizing his code the way he wants to organize it?

I spent about 24 hours helping to develop the Gromacs topology reader, and I think the overall code base has a good design and is well thought-out (in my opinion).

**Comment by swails** [comment]
> If they are interdependent....well, that sounds more like a design flaw?

You grant too much credit in suggesting this split is a byproduct of design :).  The interdependency is conceptual only -- the actual code dependency is one-way: `ParmedTools` uses `chemistry`.  `chemistry` is the set of "core" classes.  `ParmedTools` contains a set of tools that `parmed.py` uses in its program to implement high-level modifications and manipulations commonly desired of topologies (and that can be imported into your own Python script and used the same way).  They are two parts of the same library that I unfortunately split into two namespaces very early on.

**Comment by jchodera** [comment]
It may not be a problem.  This is just the first package I have encountered where there was no obvious logical connection between conda package name and Python package name.  It would reduce confusion (and conform to the "principle of least surprise") to have a one to one correspondence between conda and Python package names.  Non cyclic dependencies could still be enforced if multiple conda packages were used.

**Comment by jchodera** [comment]
All I am intending to say is that it is surprising that there is not a one to one correspondence between conda and Python package names, and that it may be less confusing if there was a one to one correspondence.  I don't know of any other packages that have different conda and Python package names (though I might simply be ignorant of these).

**Comment by swails** [comment]
> Just butting in here, but I realized fairly quickly (as I'm sure you did) that I needed to import chemistry for the functionality I needed rather than ParmEd. What's wrong with Jason organizing his code the way he wants to organize it?
> 
> I spent about 24 hours helping to develop the Gromacs topology reader, and I think the overall code base has a good design and is well thought-out (in my opinion).

I appreciate the kind words.  I did *try* to add and enforce design (logical to me) to the package as it matured (particularly the `chemistry` package), and I'm glad you found it logical as well.  I do agree that the naming scheme is suboptimal, though.  As a less obnoxious example than ParmEd, I've even been confused with scikit-learn, since the package name (scikit-learn) doesn't match the top-level import (sklearn).  It's not that I can't figure it out, or that it's complicated, but I can never remember if I should do `conda install sklearn` or `conda install scikit-learn` or what.  It's very minor and wastes very little time, but still a little bit of a wart to me.

If I got a second chance, `chemistry` would be renamed `parmed` and `ParmedTools` would become the `parmed.tools` subpackage.  The current scheme I find to be only a minor irritant, but the longer it goes on the harder it will become to change.

**Comment by davidlmobley** [comment]
I responded already in the other thread, but I think a key point is this
one:

>If I got a second chance, chemistry would be renamed parmed and ParmedTools would
become the parmed.toolssubpackage. The current scheme I find to be only a
minor irritant, but the longer it goes on the harder it will become to
change.

I think you want to decide what you want long-term, now, and move to it as
soon as possible. Otherwise you're gonna get stuck. Right now it seems like
the number of people this would affect is fairly small and people can make
the necessary changes in just a few minutes. I'm betting that within a
couple of months this is not going to be the case - I've been telling other
people about this effort and encouraging them to get involved, and already
there is enough here that it's powerful and useful. Even the "how should I
organize things?" decision will get harder to make once more people are
involved.

On Fri, Jun 5, 2015 at 10:12 AM, Jason Swails <notifications@github.com>
wrote:

> Just butting in here, but I realized fairly quickly (as I'm sure you did)
> that I needed to import chemistry for the functionality I needed rather
> than ParmEd. What's wrong with Jason organizing his code the way he wants
> to organize it?
>
> I spent about 24 hours helping to develop the Gromacs topology reader, and
> I think the overall code base has a good design and is well thought-out (in
> my opinion).
>
> I appreciate the kind words. I did *try* to add and enforce design
> (logical to me) to the package as it matured (particularly the chemistry
> package), and I'm glad you found it logical as well. I do agree that the
> naming scheme is suboptimal, though. As a less obnoxious example than
> ParmEd, I've even been confused with scikit-learn, since the package name
> (scikit-learn) doesn't match the top-level import (sklearn). It's not that
> I can't figure it out, or that it's complicated, but I can never remember
> if I should do conda install sklearn or conda install scikit-learn or
> what. It's very minor and wastes very little time, but still a little bit
> of a wart to me.
>
> If I got a second chance, chemistry would be renamed parmed and
> ParmedTools would become the parmed.tools subpackage. The current scheme
> I find to be only a minor irritant, but the longer it goes on the harder it
> will become to change.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ParmEd/ParmEd/issues/169#issuecomment-109305368>.
>



-- 
David Mobley
dmobley@gmail.com
949-385-2436

**Comment by hainm** [comment]
I totally agree with @davidlmobley. It's not too late to make huge change now. In the past, I think ParmEd is used internally in Amber without exposing much to user (only `parmed.py` is exposed).

**Comment by swails** [comment]
I don't like making ultimatums or absolute statements, since the future is so unpredictable.  But my mindset in this regard is "make the change for ParmEd 2.0 or forever hold my peace".  I initially resisted the change because of all my scripts I have depending on it and how long it had been the way it is (ca. 2011).

It took me a little while after @kyleabeauchamp proposed #85 for me to really like the alternative parmed-based package naming scheme more than `chemistry` by enough to warrant putting myself through the trouble of breaking all my scripts.  But with the idea of making `ParmedTools` the `parmed.tools` subpackage and moving *all* of the installed files into a single eponymous package, the added work I'm putting on myself seems justified.

But I also have to treat ParmEd like a package that others are beginning to use, and I can't keep pulling the carpet out from under them (or it will revert to a pet project that only I use :)).  In reality, I think @hainm is right, and that the current user base is small enough that the churn and inconvenience will be minimal (for such a large change).

And if the current users either strongly want, or don't care about, the change, then it's worth doing (if it's done now).

**Comment by davidlmobley** [comment]
@swails - yes, I agree with this completely. AND I would add that several
of us have complained about the same mental inconsistency when installing -
"wait, I installed parmed - how do I import it?" - and that's a substantial
fraction of those of us who chimed in here. I agree it's justified, worth
doing, and for 2.0.

On Fri, Jun 5, 2015 at 12:03 PM, Jason Swails <notifications@github.com>
wrote:

> I don't like making ultimatums or absolute statements, since the future is
> so unpredictable. But my mindset in this regard is "make the change for
> ParmEd 2.0 or forever hold my peace". I initially resisted the change
> because of all my scripts I have depending on it and how long it had been
> the way it is (ca. 2011).
>
> It took me a little while after @kyleabeauchamp
> <https://github.com/kyleabeauchamp> proposed #85
> <https://github.com/ParmEd/ParmEd/issues/85> for me to really like the
> alternative parmed-based package naming scheme more than chemistry by
> enough to warrant putting myself through the trouble of breaking all my
> scripts. But with the idea of making ParmedTools the parmed.tools
> subpackage and moving *all* of the installed files into a single
> eponymous package, the added work I'm putting on myself seems justified.
>
> But I also have to treat ParmEd like a package that others are beginning
> to use, and I can't keep pulling the carpet out from under them (or it will
> revert to a pet project that only I use :)). In reality, I think @hainm
> <https://github.com/hainm> is right, and that the current user base is
> small enough that the churn and inconvenience will be minimal (for such a
> large change).
>
> And if the current users either strongly want, or don't care about, the
> change, then it's worth doing (if it's done now).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ParmEd/ParmEd/issues/169#issuecomment-109341935>.
>



-- 
David Mobley
dmobley@gmail.com
949-385-2436

**Comment by hainm** [comment]
I had the same situation. I first knew `ParmEd` when Jason implemented `mbondi3` radii, but I got confused with "I installed ParmEd but why I need to import `chemistry".

**Comment by swails** [comment]
@hainm -- `mbondi3` was originally implemented in `parmed.py`, not `chemistry`, and was only in `parmed.py` ;).  Anyway...

Right now I'm leaning heavily toward making the change, mainly because the only strong opinions have been in favor of the change (and strengthened by the fact that it's now my preference, too).  I think the only established uses of ParmEd right now are `parmed.py`, support in pytraj (which @hainm can quite easily update, since his desk is shouting distance from mine) and some very small roles in FESetup's Amber support (Hannes uses Sire to do the prmtop parsing).

The way I'd like to do this is through another 2.0 beta, so I'm pinging @ChayaSt as a heads-up that when I do make this change, any Travis projects using the conda package will need to be updated.  I'm too deep in #170 right now to do this before that gets merged, but the upside is that #170 is almost done.  There should be plenty of time for that beta to get vetted before OMM 6.3 is released.

**Comment by hainm** [comment]
oops, so what's the relationship between `parmed.py` and `parmed` :D and `ParmEd`. :D

**Comment by swails** [comment]
This is not that ambiguous:

- ParmEd is the name of the program and library.
- `parmed.py` is the driver program (it can be converted to an entry point defined in `parmed/__main__.py`)
- `parmed` is the library package -- same as the project name with a PEP8-ized package name.

**Comment by hainm** [comment]
I see.

**Comment by jchodera** [comment]
It may be useful to call the console script `parmed` what you call the "driver program") as well if it is just going to call the `parmed/__main__.py` entry point.

**Comment by swails** [comment]
> I don't know of any other packages that have different conda and Python package names (though I might simply be ignorant of these).

You do know of one.  We all do:

```
conda install openmm
import simtk
```

Just sayin ;).

**Comment by jchodera** [comment]
Good point!  Also weird.

**Comment by swails** [comment]
OK, most of these issues have been worked out.  As far as I can tell, the only real barrier to an official ParmEd release is waiting on OpenMM 6.3 to come out.  I've had a number of Beta releases, and @davidlmobley has helped work out a lot of kinks.  So I think the conda packages are working well, and we are hopefully approaching some semblance of stability.

I'm also done with the main API changes at this point (libraries and scripts are renamed, coordinate handling is improved and made consistent across all classes, etc.).

**Comment by leeping** [comment]
I like the idea of doing a 2.0 release.  Will you also update ParmEd in AmberTools and change `chemistry -> parmed` where AmberTools imports ParmEd functions (in `pysander` for example)?

**Comment by jchodera** [comment]
Presumably, OpenMM 6.3 could be released on Mon 6 Jul if there were no issues noticed?

Note that the `openmm-dev` package is not being updated automatically until we find a solution for https://github.com/omnia-md/openmm-dev/issues/5 so ping me if you want me to update it manually.

**Comment by swails** [comment]
> Will you also update ParmEd in AmberTools and change chemistry -> parmed where AmberTools imports ParmEd functions (in pysander for example)?

No, the AmberTools release will stay as-is.  Luckily there won't be any conflicts, as the imports have different names.

In general, Amber has not historically required Python, so our users don't have to use Python in any aspect of their work (many of the developers actually don't make any use of Python).  And the current version has some rather jarring differences compared to the AmberTools 15 version.  In particular, I've dropped support for Python 2.4 - 2.6 and started requiring numpy.

While I don't think this is a bad thing (as I've mentioned in other issues), I don't want what should be a small update to all of a sudden break things for a lot of people.

**Comment by swails** [comment]
> Presumably, OpenMM 6.3 could be released on Mon 6 Jul if there were no issues noticed?

No idea.  We'll see what Peter does here.  Once that happens, I'll start to move things forward for an actual release.

**Comment by jchodera** [comment]
Looks like the binstar version is still `2.0beta3`.  Does it make sense to have an official 2.0 now?

**Comment by davidlmobley** [comment]
Yes, I think so. (I had a note to myself to ask about updating this but
hadn't gotten to it yet).

On Wed, Aug 19, 2015 at 1:17 PM, John Chodera <notifications@github.com>
wrote:

> Looks like the binstar version is still 2.0beta3. Does it make sense to
> have an official 2.0 now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ParmEd/ParmEd/issues/169#issuecomment-132767993>.
>



-- 
David Mobley
dmobley@gmail.com
949-385-2436

**Comment by swails** [comment]
OpenMM 6.3 has been released, so I think we can get ready to cut an official release of ParmEd

**Comment by davidlmobley** [comment]
Agreed.

**Comment by swails** [comment]
Done.

**Comment by jchodera** [comment]
Woot.
